### PR TITLE
Add example for `FunctionFactory`

### LINKED ISSUE #9448: Add example for `CREATE FUNCTION` & `FunctionFactory`
### Is your feature request related to a problem or challenge?

Once we get #9333 merged create an example to demonstrate how to use it 

### Describe the solution you'd like

It would be great if we can wait for #9304 to get merged as well, it would make example easier to implement

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by milenkovicm**
## Which issue does this PR close?

Closes #9448.

## Rationale for this change

Explained in #9448. It covers example for functionality added in #9304 & #9333

## What changes are included in this PR?

Example demonstrating functionality

## Are these changes tested?

Yes 

## Are there any user-facing changes?

No

### CODE REVIEW

**Comment by milenkovicm** [comment]
@alamb when you get chance 😀 thanks ! 
Do we need more documentation on it ?

**Comment by alamb** [comment]
Thanks @milenkovicm  -- I will review this over the next day or two. I am sorry I am behind on reviews and catching up now

**Review by alamb — APPROVED** [review]
Thank you so much @milenkovicm  -- this feature is *SOO* cool. 

Can you just imagine a library of `FunctionHandlers` so people could pick from a list of types of functions they wanted to support:
* I want support for pytorch models"
* I want to support SQL macros
* Oh, and also I want to support WASM 

A few imports, a `cargo build` command and off they go. 🤩 

I pushed some word smithing comments up to your branch, as well as some comments that are more in the "explain it line by line" style -- let me know what you think

We can revert them if you prefer. What you had was great and I just got excited as I had it checked out locally

File: datafusion-examples/README.md [hunk 1]
```diff
@@ -54,6 +54,7 @@ cargo run --example csv_sql
 - [`deserialize_to_struct.rs`](examples/deserialize_to_struct.rs): Convert query results into rust structs using serde
 - [`expr_api.rs`](examples/expr_api.rs): Create, execute, simplify and analyze `Expr`s
 - [`flight_sql_server.rs`](examples/flight/flight_sql_server.rs): Run DataFusion as a standalone process and execute SQL queries from JDBC clients
+- [`function_factory.rs`](examples/function_factory.rs): Register `CREATE FUNCTION` handler to implement SQL macros
```
Comment (alamb) [inline datafusion-examples/README.md:57]: I added the example

File: datafusion-examples/examples/function_factory.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    // argument
+    let sql = r#"
+    CREATE FUNCTION f2(BIGINT, BIGINT)
+        RETURNS BIGINT
+        RETURN $1 + f1($2)
+    "#;
+
+    ctx.sql(sql).await?.show().await?;
+
+    // Invoke f2, and we expect to see 1 + (1 + 2) = 4
```
Comment (alamb) [inline datafusion-examples/examples/function_factory.rs:69]: I removed some of the other things that didn't really help teach people how to use this feature any more (they were more like ensuring all features got covered) to keep it shorter

File: datafusion-examples/examples/function_factory.rs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    fn invoke(
+        &self,
+        _args: &[datafusion_expr::ColumnarValue],
+    ) -> Result<datafusion_expr::ColumnarValue> {
+        // Since this function is always simplified to another expression, it
+        // should never actually be invoked
+        internal_err!("This function should not get invoked!")
+    }
+
+    /// The simplify function is called to simply a call such as `f2(2)`. This
```
Comment (alamb) [inline datafusion-examples/examples/function_factory.rs:142]: I tried to add some comments inline that explained what the functions were doing with more words, but hopefully for someone who was seeing if this worked for their usecase

**Comment by milenkovicm** [comment]
Thanks a lot @alamb, your additions totally make sense.
One question for you, is there a plan to add `simplify` to other function types? 

Also, inspired by simplification you've done in `SessionContext`, maybe we should add companion object to `FunctionFactory` so it can register configuration as well. At the moment its a bit of work to register configuration:

https://github.com/milenkovicm/torchfusion/blob/634c2a4e39b81e1969671db8faea50b96bc43c7f/src/lib.rs#L101

```rust
let mut session_config = SessionConfig::new().with_information_schema(true);

session_config
      .options_mut()
      .extensions
      // register torch factory configuration
      .insert(TorchConfig::default());
```

it would be much better if config could be registered when `ctx.with_function_factory(...)` called ... but that could be a follow up PR

**Comment by alamb** [comment]
> One question for you, is there a plan to add simplify to other function types?

Excellent idea -- I have filed
* https://github.com/apache/arrow-datafusion/issues/9526
* https://github.com/apache/arrow-datafusion/issues/9527

**Comment by alamb** [comment]
Thanks again @milenkovicm

**Pull request merged**